### PR TITLE
Implement reboot functionality

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: src/bsp/reboot.c [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#include "reboot.h"
```
Comment (al1img) [inline src/bsp/reboot.c:1]: Local header should last one.

File: src/bsp/reboot.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    if (rc)
+    {
+        printk("failed to reboot system (%d)\n", rc);
+    }
+}
+
+int start_reboot_monitoring()
+{
+    int ret;
+    int intervalMs = CONFIG_AOS_REBOOT_CHECKING_TIMEOUT;
```
Comment (al1img) [inline src/bsp/reboot.c:48]: Add define in header file instead of config.

File: src/main.cpp [hunk 3]
```diff
@@ -35,6 +39,12 @@ int main(void)
     }
 #endif
 
+#ifdef CONFIG_AOS_REBOOT
+    if (auto ret = start_reboot_monitoring()) {
+        printk("failed to start reboot monitoring (%d)\n", ret);
```
Comment (al1img) [inline src/main.cpp:44]: Start with capital letter. Also fix for `create_domains`.

File: src/main.cpp [hunk 4]
```diff
@@ -48,5 +58,11 @@ int main(void)
     auto err = app.Init();
     __ASSERT(err.IsNone(), "Error initializing application: %s", err.Message());
 
+#ifdef CONFIG_AOS_REBOOT
+    if (auto ret = stop_reboot_monitoring()) {
+        printk("failed to stop reboot monitoring (%d)\n", ret);
```
Comment (al1img) [inline src/main.cpp:63]: Ditto.

File: Kconfig [hunk 5]
```diff
@@ -17,6 +17,22 @@ config AOS_SM_VCHAN_RX_PATH
...
+config AOS_REBOOT_XEN_STORE_VALUE_LEN
+	int "Length reboot value"
+	default 10
+
+config AOS_REBOOT_CHECKING_TIMEOUT
+	int "Timeout for checking reboot"
+	default 10000
+
+config AOS_REBOOT
+	bool "Custom reboot"
```
Comment (al1img) [inline Kconfig:33]: "Reboot by request from guest domain"

File: src/main.cpp [hunk 6]
```diff
@@ -35,6 +39,12 @@ int main(void)
     }
 #endif
 
+#ifdef CONFIG_AOS_REBOOT
```
Comment (al1img) [inline src/main.cpp:42]: Let's have it always for not native applications.

File: src/bsp/reboot.h [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#ifndef REBOOT_H_
+#define REBOOT_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+int start_reboot_monitoring();
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/bsp/reboot.h]: Done

File: src/bsp/reboot.c (hunk 1 above)
Comment (MykolaSolyanko) [reply src/bsp/reboot.c]: Done

File: src/bsp/reboot.c (hunk 2 above)
Comment (MykolaSolyanko) [reply src/bsp/reboot.c]: Done

File: src/main.cpp (hunk 3 above)
Comment (MykolaSolyanko) [reply src/main.cpp]: Done

File: src/main.cpp (hunk 4 above)
Comment (MykolaSolyanko) [reply src/main.cpp]: Done

File: Kconfig (hunk 5 above)
Comment (MykolaSolyanko) [reply Kconfig]: Removed this config

File: src/main.cpp (hunk 6 above)
Comment (MykolaSolyanko) [reply src/main.cpp]: Done

File: Kconfig [hunk 8]
```diff
@@ -17,6 +17,10 @@ config AOS_SM_VCHAN_RX_PATH
 	string "Path to the SM vchan"
 	default "/local/domain/1/tmp/vchan/sm_rx"
 
+config AOS_REBOOT_XEN_STORE_PATH
+	string "Path reboot"
+	default "/dev/xen/xenbus/control/user-reboot"
```
Comment (al1img) [inline Kconfig:22]: It should be something similar to vchan path: "/local/domain/1/control/user-reboot". On linux domain it also possible to use relative path: "control/user-reboot". Could you check?

File: CMakeLists.txt [hunk 9]
```diff
@@ -93,8 +93,8 @@ target_sources_ifdef(
 # NOTE: Include source files for qemu boards.
 target_sources_ifdef(CONFIG_BOARD_QEMU_CORTEX_A53 app PRIVATE src/domains/domd/domd_cfg_xenvm.c)
 
-# NOTE: Include source files for LFS on MMC support
-target_sources_ifdef(CONFIG_AOS_MOUNT_LFS_MMC app PRIVATE src/bsp/mount.c)
+# NOTE: Include source files for custom system reboot
```
Comment (al1img) [inline CMakeLists.txt:96]: Fix comment.

File: src/bsp/reboot.h [hunk 10]
```diff
@@ -0,0 +1,17 @@
+#ifndef REBOOT_H_
+#define REBOOT_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define CONFIG_AOS_REBOOT_CHECKING_TIMEOUT 10000
```
Comment (al1img) [inline src/bsp/reboot.h:8]: timeout -> period, interval,  also add unit suffix `MS`.

File: src/bsp/reboot.c [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+struct k_work  perform_reboot_work;
+
+void perform_system_reboot(struct k_work* work)
+{
+    int                rc, value;
+    static const char* path = CONFIG_AOS_REBOOT_XEN_STORE_PATH;
+
+    rc = xss_read_integer(path, &value);
+    if (rc) {
+        return;
```
Comment (al1img) [inline src/bsp/reboot.c:17]: Should we put some error here?

File: src/bsp/reboot.h [hunk 12]
```diff
@@ -0,0 +1,16 @@
+#ifndef REBOOT_H_
+#define REBOOT_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define CONFIG_AOS_REBOOT_CHECKING_TIMEOUT_SEC 10
```
Comment (al1img) [inline src/bsp/reboot.h]: timeout -> period, interval

File: src/bsp/reboot.c [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+static void trigger_reboot_check(struct k_timer* timer_id)
+{
+    k_work_submit(&perform_reboot_work);
+}
+
+void monitor_reboot_condition()
+{
+    printk("******Start monitoring for reboot conditions*******\n");
```
Comment (al1img) [inline src/bsp/reboot.c:40]: Remove "*****"
"Start guest reboot watcher."

File: Kconfig (hunk 8 above)
Comment (MykolaSolyanko) [reply Kconfig]: Done, use the full path

File: CMakeLists.txt (hunk 9 above)
Comment (MykolaSolyanko) [reply CMakeLists.txt]: Done

File: Kconfig [hunk 14]
```diff
@@ -17,6 +17,10 @@ config AOS_SM_VCHAN_RX_PATH
 	string "Path to the SM vchan"
 	default "/local/domain/1/tmp/vchan/sm_rx"
 
+config AOS_REBOOT_XEN_STORE_PATH
+	string "Path reboot"
```
Comment (al1img) [inline Kconfig:21]: Path to user reboot request.

File: prj.conf [hunk 15]
```diff
@@ -131,6 +131,10 @@ CONFIG_XEN_SHELL=y
 CONFIG_XRUN=y
 CONFIG_XRUN_SHELL_CMDS=y
 
+# Enable cpu power management
+
+CONFIG_PM_CPU_OPS=y
```
Comment (al1img) [inline prj.conf:136]: Move to system components section. It is not related to xen.

File: CMakeLists.txt [hunk 16]
```diff
@@ -93,8 +93,8 @@ target_sources_ifdef(
 # NOTE: Include source files for qemu boards.
 target_sources_ifdef(CONFIG_BOARD_QEMU_CORTEX_A53 app PRIVATE src/domains/domd/domd_cfg_xenvm.c)
 
-# NOTE: Include source files for LFS on MMC support
-target_sources_ifdef(CONFIG_AOS_MOUNT_LFS_MMC app PRIVATE src/bsp/mount.c)
+# NOTE: Include source files for custom system operations
```
Comment (al1img) [inline CMakeLists.txt:96]: Include HW-specific source file.

File: src/bsp/reboot.h [hunk 17]
```diff
@@ -0,0 +1,16 @@
+#ifndef REBOOT_H_
+#define REBOOT_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define CONFIG_AOS_REBOOT_CHECKING_PERIOD_SEC 10
+
+void monitor_reboot_condition();
```
Comment (al1img) [inline src/bsp/reboot.h:10]: reboot_watcher_init

File: src/bsp/reboot.h (hunk 12 above)
Comment (MykolaSolyanko) [reply src/bsp/reboot.h]: Done

File: src/bsp/reboot.c (hunk 11 above)
Comment (MykolaSolyanko) [reply src/bsp/reboot.c:17]: in this case, we will receive error every time until someone write into this path

File: src/bsp/reboot.c (hunk 13 above)
Comment (MykolaSolyanko) [reply src/bsp/reboot.c:40]: Done

File: prj.conf (hunk 15 above)
Comment (MykolaSolyanko) [reply prj.conf]: Done

File: CMakeLists.txt (hunk 16 above)
Comment (MykolaSolyanko) [reply CMakeLists.txt]: Done

File: src/bsp/reboot.h (hunk 17 above)
Comment (MykolaSolyanko) [reply src/bsp/reboot.h]: Done

**Review by al1img — APPROVED** [review]

**Review by al1img — APPROVED** [review]

**Pull request merged**